qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,273 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""A completion model backed by SQL tables."""
+
+import re
+
+from PyQt5.QtCore import Qt, QModelIndex, QAbstractItemModel
+from PyQt5.QtSql import QSqlQueryModel
+
+from qutebrowser.utils import log
+from qutebrowser.misc import sql
+
+
+class _SqlCompletionCategory(QSqlQueryModel):
+    def __init__(self, name, sort_by, sort_order, limit, select, where,
+                 columns_to_filter, parent=None):
+        super().__init__(parent=parent)
+        self.tablename = name
+
+        query = sql.run_query('select * from {} limit 1'.format(name))
+        self._fields = [query.record().fieldName(i) for i in columns_to_filter]
+
+        querystr = 'select {} from {} where ('.format(select, name)
+        querystr += ' or '.join("{} like ? escape '\\'".format(f)
+                                for f in self._fields)
+        querystr += ')'
+        if where:
+            querystr += ' and ' + where
+
+        if sort_by:
+            sortstr = 'asc' if sort_order == Qt.AscendingOrder else 'desc'
+            querystr += ' order by {} {}'.format(sort_by, sortstr)
```

Comment:
Maybe assert that `sort_order` is `None` if `sort_by` is None to avoid mistakes?